cache-prefetch.go

```diff
@@ -0,0 +1,221 @@
+package rdns
+
+import (
+	"errors"
+	"expvar"
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+	"strings"
+	"sync"
+	"time"
+)
+
+type CachePrefetch struct {
+	CachePrefetchOptions
+	id       string
+	resolver Resolver
+	mu       sync.Mutex
+	metrics  *CachePrefetchMetrics
+}
+
+type CachePrefetchMetrics struct {
+	// Cache hit count.
+	domainEntries map[string]CachePrefetchEntry
+}
+
+// 0 for no prefetching job
+// 1 for prefetching job is active
+// 2 for stopped by error
+type PrefetchState int
+
+const (
+	PrefetchStateNone   = iota
+	PrefetchStateActive = 1
+	PrefetchStateOther  = 2
+)
+
+type CachePrefetchEntry struct {
+	// request hit count.
+	hit expvar.Int
+
+	prefetchState PrefetchState
+	// store the time to live
+	msg *dns.Msg
```

Comment:
Unfortunately, this does make the cache double up. Is there a better way to cache a DNS question? I tried just storing the DNS question which would result in a `SERVFAIL` is there more I need to store or less?